### PR TITLE
Eliminating var-like setter tpe_= on Tree.

### PR DESCRIPTION
**Pull request by paulp**
Deprecated tpe_= on Tree, which is redundant with and
less useful than setType. To provide a small layer of
insulation from the direct nulling out of mutable fields
used to signal the typer, added def clearType() which is
merely tree.tpe = null but is shamefaced about the null
and var-settings parts like a respectable method should be.

### CODE REVIEW

**Comment by paulp** [comment]
Stop me before I drift again! Review by @adriaanm

**Comment by adriaanm** [comment]
i'll look at it now, but it already has a sad-pullrequest-greybar

**Comment by paulp** [comment]
Yes, I'll ward off the Grays, but I'd rather not go into a rebasing loop pining for a reviewer.

File: src/compiler/scala/tools/nsc/transform/Erasure.scala [hunk 1]
```diff
@@ -1260,13 +1259,12 @@ abstract class Erasure extends AddInterfaces
...
             case ArrayValue(elemtpt, trees) =>
               treeCopy.ArrayValue(
-                tree1, elemtpt setType specialScalaErasure.applyInArray(elemtpt.tpe), trees map transform) setType null
+                tree1, elemtpt setType specialScalaErasure.applyInArray(elemtpt.tpe), trees map transform).clearType()
             case DefDef(_, _, _, _, tpt, _) =>
-              val result = super.transform(tree1) setType null
-              tpt.tpe = specialErasure(tree1.symbol)(tree1.symbol.tpe).resultType
-              result
+              try super.transform(tree1).clearType()
+              finally tpt setType specialErasure(tree1.symbol)(tree1.symbol.tpe).resultType
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: cunning -- does the backend have matching smarts?

Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: Matching? It's not a coordinated maneuver, it's just this transform:

```
val x = expr ; sideEffect() ; x
```

into

```
try expr finally sideEffect()
```

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: yes, i'm just hoping the backend then doesn't go all exception-y on us
but instead produces bytecode equivalent to the original statement
 (i don't know, just never checked the bytecode we generate for this pattern before)

Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: Oh. Well it's a reasonable thing to worry about, I'll check.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: Thanks. Everything else LGTM. Especially the `modifyType`s are a clear win.

Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: The modifyTypes are a clear win for clearity at least.

The backend is not the undiscovered genius we might have hoped. The finalizer gets duplicated. I don't think it bears sweating (it's only one case in one transformer) but I can pull it if you want.

```
@@ -88,14 +88,21 @@ public class T extends scala.reflect.api.Trees$Transformer {
       : astore
       : aload
       : invokespecial #                  // Method scala/reflect/api/Trees$Transformer.transform:(Lscala/reflect/api/Trees$TreeApi;)Lscala/reflect/api/Trees$TreeApi;
-      : astore
       : aload
       : invokevirtual #                  // Method bippy:(Lscala/reflect/api/Trees$TreeApi;)Lscala/reflect/api/Trees$TreeApi;
       : pop
-      : aload
       : astore
       : aload
       : areturn
+      : astore
+      : aload
+      : invokevirtual #                  // Method bippy:(Lscala/reflect/api/Trees$TreeApi;)Lscala/reflect/api/Trees$TreeApi;
+      : pop
+      : aload
+      : athrow
```

**Comment by paulp** [comment]
PLS REBUILD ALL

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 1 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: that sounds acceptable (and yes, I thought all of this was purely about clarity)

**Comment by adriaanm** [comment]
LGTM

File: src/compiler/scala/tools/nsc/transform/Erasure.scala (hunk 1 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/Erasure.scala:1265]: Feel the calm, pure clarity washing over you.

**Pull request merged**